src/TokenHandler/AccountManager.mo

```diff
@@ -0,0 +1,408 @@
+    Nat, // totalWithdrawn_
+    Nat, // depositedFunds_
+    Nat, // queuedFunds
+  );
+
+  /// Manages accounts and funds for users.
+  /// Handles deposit, withdrawal, and consolidation operations.
+  public class AccountManager(
+    icrc1LedgerPrincipal : Principal,
+    ownPrincipal : Principal,
+    journal : Journal.Journal,
+    isFrozen : () -> Bool,
+    freezeCallback : (text : Text) -> (),
+    creditRegistry : CreditRegistry.CreditRegistry,
+  ) {
+
+    let icrc1Ledger = actor (Principal.toText(icrc1LedgerPrincipal)) : ICRC1.ICRC1Ledger;
+
+    /// Manages deposit balances for each user.
+    let depositRegistry : DepositRegistry.DepositRegistry = DepositRegistry.DepositRegistry(freezeCallback);
+
+    /// Manages acklog of principals waiting for processing.
+    let backlog : Backlog.Backlog = Backlog.Backlog();
+
+    /// Current fee amount.
+    var fee_ : Nat = 0;
+
+    /// Total deposited funds across all accounts.
+    var depositedFunds_ : Nat = 0;
+
+    /// Total amount consolidated. Accumulated value.
+    var totalConsolidated_ : Nat = 0;
+
+    /// Total amount withdrawn. Accumulated value.
+    var totalWithdrawn_ : Nat = 0;
+
+    /// Total funds queued for consolidation.
+    var queuedFunds : Nat = 0;
+
+    /// Total funds underway for consolidation.
+    var underwayFunds : Nat = 0;
+
+    /// Retrieves the current fee amount.
+    public func fee() : Nat = fee_;
+
+    /// Updates the fee amount based on the ICRC1 ledger.
+    public func updateFee() : async* Nat {
+      let newFee = await icrc1Ledger.icrc1_fee();
+      if (fee_ != newFee) {
+        journal.push((Time.now(), ownPrincipal, #feeUpdated({ old = fee_; new = newFee })));
+        fee_ := newFee;
+        recalculateBacklog();
+      };
+      newFee;
+    };
+
+    /// Retrieves the sum of all current deposits. This value is nearly the same as backlogFunds(), but includes
+    /// entries, which could not be added to backlog, for instance when balance less than fee.
+    /// It's always >= backlogFunds()
+    public func depositedFunds() : Nat = depositedFunds_;
+
+    /// Retrieves the sum of all successful consolidations.
+    public func totalConsolidated() : Nat = totalConsolidated_;
+
+    /// Retrieves the sum of all deductions from the main account.
+    public func totalWithdrawn() : Nat = totalWithdrawn_;
+
+    /// Retrieves the calculated balance of the main account.
+    public func consolidatedFunds() : Nat = totalConsolidated_ - totalWithdrawn_;
+
+    /// Returns the size of the consolidation backlog.
+    public func backlogSize() : Nat = backlog.size();
+
+    /// Returns the sum of all deposits in the backlog.
+    public func backlogFunds() : Nat = queuedFunds + underwayFunds;
+
+    /// Notifies of a deposit and schedules backlog processing.
+    /// Returns the newly detected deposit if successful.
+    public func notify(p : Principal) : async* ?Nat {
+      if (isFrozen() or not depositRegistry.lock(p)) return null;
```

Comment:
Do we need the `isFrozen()` part here? AccountManager is only a class, not an actor, so it has surrounding code. We call the freezeCallback() to inform the surrounding code. So the surrounding code could catch the frozen state and block calling AccountManager.notify().

Reply:
Yes, you are right.
We can move the check to TokenHandler. Or TokenHandler also should not contain it and should just inform the client code about being frozen?

Reply:
I think TokenHandler as a whole can reject calls when it is frozen, so that it indirectly informs the client code that it is frozen. We don't need a callback to inform the client code. 

If the TokenHandler is bug-free then it won't freeze. So normal client code should not be bothered with freezing logic.

Reply:
Didn't manage to move the check to a higher level. We need to know if the handler is frozen inside methods such as `credit()`, `debit()`, `processBacklog()`. So we have to break the abstraction a little with passing `isFrozen()`

What do you think?